File: 计算签名/HttpDemo.py
import base64
import hashlib
import hmac
import json
import random
import string
import time
from datetime import datetime, timedelta

import requests

# ali config
"""
api_link: https://developer.umeng.com/docs/143070/detail/144783
signature_link: https://developer.umeng.com/docs/143070/detail/144785
"""


def generate_code(length):
    key = string.digits + 'abcdef'
    str_list = random.sample(key + key, length)  # length >= len(key+key)
    result_str = ''.join(str_list)
    return result_str


# aliyun config
ali_appkey = 'xxxxxxxx'
ali_app_secret = 'xxxxxxxx'

# umeng config
host_url = 'https://verify5.market.alicloudapi.com'
path_url = '/api/v1/mobile/info'
umeng_app_key = 'xxxxxxxx'

token = 'xxxxxxxx'

body_data_str = json.dumps({'token': token})
content_md5_str = base64.b64encode(hashlib.md5(body_data_str.encode("UTF-8")).digest()).decode("UTF-8")

now_date_time = datetime.utcnow() + timedelta(hours=8)
timestamp = str(int(time.mktime(now_date_time.timetuple()) * 1000))

sign_headers_dict = {
    'X-Ca-Key': ali_appkey,
    'X-Ca-Stage': 'RELEASE',
    'X-Ca-Timestamp': timestamp,
    'X-Ca-Version': '1'
}

sign_headers_str = ''
for key in sorted(sign_headers_dict.keys()):
    value = sign_headers_dict[key]
    sign_headers_str = sign_headers_str + key + ':' + value + '\n'

sign_dict = {
    'HTTPMethod': 'POST',
    'Accept': 'application/json',
    'Content-MD5': content_md5_str,
    'Content-Type': 'application/json; charset=UTF-8',
    'Date': None,
    'Headers': sign_headers_str,
    'Url': path_url + '?' + 'appkey=' + umeng_app_key,
}

string_to_sign = ""
for key in sign_dict.keys():
    value = sign_dict[key]
    if key not in ['Headers', 'Url']:
        if value:
            string_to_sign = string_to_sign + value + "\n"
        else:
            string_to_sign = string_to_sign + "\n"
    else:
        string_to_sign = string_to_sign + value

key_bytes = ali_app_secret.encode('utf-8')
text_bytes = string_to_sign.encode('utf-8')
hash_bytes = hmac.new(key_bytes, text_bytes, hashlib.sha256).digest()
signature_str = base64.b64encode(hash_bytes).decode('utf-8')

nonce_str = generate_code(8) + '-' + generate_code(4) + '-' + generate_code(4) + '-' + generate_code(
    4) + '-' + generate_code(12)

headers_dict = {
    'Content-Type': 'application/json; charset=UTF-8',
    'Accept': 'application/json',
    'X-Ca-Version': '1',
    'X-Ca-Signature-Headers': 'X-Ca-Version,X-Ca-Stage,X-Ca-Key,X-Ca-Timestamp',
    'X-Ca-Stage': 'RELEASE',
    'X-Ca-Key': ali_appkey,
    'X-Ca-Timestamp': timestamp,
    'X-Ca-Nonce': nonce_str,
    'Content-MD5': content_md5_str,
    'X-Ca-Signature': signature_str
}

request_url = host_url + path_url + '?' + 'appkey=' + umeng_app_key
response = requests.post(url=request_url, data=body_data_str.encode("UTF-8"), headers=headers_dict)

print(response)
print(response.headers)
print(response.text)

File: 计算签名/HttpDemo.go
package main

import (
	"crypto/hmac"
	"crypto/md5"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
	"time"
)

func generate_md5(content string) string {
	b := md5.Sum([]byte(content))
	return base64.StdEncoding.EncodeToString(b[:])
}

func generate_sign(method string, req *http.Request, aliyunAppSecret string, uri string) string {
	var sign strings.Builder
	sign.WriteString(method + "\n")
	sign.WriteString(req.Header.Get("Accept") + "\n")
	sign.WriteString(req.Header.Get("Content-MD5") + "\n")
	sign.WriteString(req.Header.Get("Content-Type") + "\n")
	sign.WriteString(req.Header.Get("Date") + "\n")
	sign.WriteString("X-Ca-Key:" + req.Header.Get("X-Ca-Key") + "\n")
	sign.WriteString("X-Ca-Nonce:" + req.Header.Get("X-Ca-Nonce") + "\n")
	sign.WriteString("X-Ca-Stage:" + req.Header.Get("X-Ca-Stage") + "\n")
	sign.WriteString("X-Ca-Timestamp:" + req.Header.Get("X-Ca-Timestamp") + "\n")
	sign.WriteString("X-Ca-Version:" + req.Header.Get("X-Ca-Version") + "\n")
	sign.WriteString(uri)
	key := []byte(aliyunAppSecret)
	h := hmac.New(sha256.New, key)
	h.Write([]byte(sign.String()))
	return base64.StdEncoding.EncodeToString(h.Sum(nil))
}

func main() {
	// 初始化参数
	umengAppkey := "123"
	aliyunAppKey := "20000000"
	aliyunAppSecret := "xxxxxxxx"
	uri := "/api/v1/mobile/info?appkey=" + umengAppkey
	url := "https://verify5.market.alicloudapi.com" + uri
	httpBody := `{"token":"1234"}`
	method := "POST"
	// 初始化http请求
	payload := strings.NewReader(httpBody)
	client := &http.Client{}
	req, err := http.NewRequest(method, url, payload)
	if err != nil {
		fmt.Println(err)
		return
	}
	// 设置header
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-MD5", generate_md5(httpBody))
	req.Header.Add("Content-Type", "application/json; charset=UTF-8")
	req.Header.Add("Date", fmt.Sprint(time.Now()))
	req.Header.Add("X-Ca-Key", aliyunAppKey)
	req.Header.Add("X-Ca-Nonce", fmt.Sprintf("%x", md5.Sum([]byte("my_salt"+fmt.Sprint(time.Now().UnixNano())))))
	req.Header.Add("X-Ca-Stage", "RELEASE")
	req.Header.Add("X-Ca-Timestamp", fmt.Sprint(time.Now().UnixNano()/1000000))
	req.Header.Add("X-Ca-Version", "1")
	req.Header.Add("X-Ca-Signature-Headers", "X-Ca-Key, X-Ca-Nonce, X-Ca-Stage, X-Ca-Timestamp, X-Ca-Version")
	req.Header.Add("X-Ca-Signature", generate_sign(method, req, aliyunAppSecret, uri))
	// 发起请求
	res, err := client.Do(req)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer res.Body.Close()
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println(res.Header)
	fmt.Println(string(body))
}
